index.js

```diff
@@ -1,174 +1,221 @@
+    Object.defineProperty(this, 'opts', readonly(opts))
+    Object.defineProperty(this, 'asBlock', readonly(this))
+  }
 
-    get codec () {
-      if (this.opts.code) {
-        this.opts.codec = multicodec.get(this.opts.code).name
+  get hasher () {
+    if (this.opts.cid) {
+      if (!this.opts.hasher) {
+        this.opts.hasher = Block.codecs.get(this.opts.cid.multihash.code)
+      } else if (this.opts.hasher.code !== this.opts.cid.multihash.code) {
+        this.opts.hasher = Block.codecs.get(this.opts.cid.multihash.code)
+      } else {
+        return this.opts.hasher || sha256
       }
-      if (this.opts.cid) {
-        this.opts.codec = multicodec.get(this.opts.cid.code).name
-      }
-      return this.opts.codec
+      if (!this.opts.hasher) throw new Error('Do not have hash implementation')
     }
+    return this.opts.hasher || sha256
+  }
+
+  source () {
+    if (this.opts.cid || this.opts.data ||
+        this._encoded || this._decoded) return null
+    if (!this.opts.source) return null
+    return this.opts.source
+  }
+
+  async cid () {
+    if (this.opts.cid) return this.opts.cid
+    const hash = await this.hasher.digest(this.encodeUnsafe())
+    const cid = CID.create(1, this.opts.codec.code, hash)
+    this.opts.cid = cid
+    // https://github.com/bcoe/c8/issues/135
+    /* c8 ignore next */
+    return cid
+  }
 
-    get code () {
-      if (this.opts.cid) return this.opts.cid.code
-      if (!this.opts.code) {
-        this.opts.code = multicodec.get(this.codec).code
+  get codec () {
+    if (this.opts.cid) {
+      if (!this.opts.codec || this.opts.codec.code !== this.opts.cid.code) {
+        this.opts.codec = Block.codecs.get(this.opts.cid.code)
       }
-      return this.opts.code
+    } else if (this.opts.code) {
+      this.opts.codec = Block.codecs.get(this.opts.code)
     }
+    return this.opts.codec.name
+  }
 
-    async validate () {
-      // if we haven't created a CID yet we know it will be valid :)
-      if (!this.opts.cid) return true
-      const cid = await this.cid()
-      const data = this.encodeUnsafe()
-      // https://github.com/bcoe/c8/issues/135
-      /* c8 ignore next */
-      return multihash.validate(cid.multihash, data)
+  get code () {
+    if (this.opts.cid) return this.opts.cid.code
+    if (!this.opts.code) {
+      this.opts.code = this.opts.codec.code
     }
+    return this.opts.code
+  }
 
-    _encode () {
-      this._encoded = this.opts.data || multicodec.get(this.code).encode(this.opts.source)
-    }
+  async validate () {
+    // if we haven't created a CID yet we know it will be valid :)
+    if (!this.opts.cid) return true
+    if (!this.opts.hasher) throw new Error('Must have hasher in order to perform comparison')
+    const cid = await this.cid()
+    const data = this.encodeUnsafe()
+    const hash = await this.hasher.digest(data)
+    if (bytes.equals(cid.multihash.bytes, hash.bytes)) return true
+    throw new Error('Bytes do not match')
```

Comment:
💭 I think this another case where dedicated constructors would make more sense. E.g. you could have a `Block.createUnsafe(cid, bytes)` and `Block.createSafe(cid, bytes)` where later delegates to former after validation. That would also get rid of this in cases where you know no validation is needed like when you create block e.g from JS value.

Reply:
I was planning on moving to dedicated constructors but hadn’t identified these as separate methods, but I like it. Although I think I’m just going to go with `create()` and `createUnsafe()` to match the other unsafe API forms.